pyresample/geometry.py

```diff
@@ -892,6 +1137,239 @@ def compute_optimal_bb_area(self, proj_dict=None):
+        The desired array output format.
+        Accepted formats are: ['Numpy','Dask', 'DataArray_Numpy','DataArray_Dask']
+    dims : tuple, optional
+        Optional argument for the specification of DataArray dimension names
+        if input array is Numpy or Dask.
+        Provide a tuple with (y_dimname, x_dimname).
+        The default is None --> (dim_0, dim_1)
+
+    Returns
+    -------
+    dst_arr : (np.ndarray, da.Array, xr.DataArray)
+        The converted 2D array.
+    src_format: str
+        The source format of the 2D array.
+
+    """
+    import dask.array as da
+    import numpy as np
+    import xarray as xr
+
+    # Checks
+    valid_format = ['Numpy', 'Dask', 'DataArray_Numpy', 'DataArray_Dask']
+    if not isinstance(to, str):
+        raise TypeError("'to' must be a string indicating the conversion array format.")
+    if not np.isin(to.lower(), np.char.lower(valid_format)):
+        raise ValueError("Valid conversion array formats are {}".format(valid_format))
+    if not isinstance(arr, (np.ndarray, da.Array, xr.DataArray)):
+        raise TypeError("The provided array must be either a np.ndarray, a dask.Array or a xr.DataArray.")
+    # Numpy
+    if isinstance(arr, np.ndarray):
+        if to.lower() == 'numpy':
+            dst_arr = arr
+        elif to.lower() == 'dask':
+            dst_arr = da.from_array(arr)
+        elif to.lower() == 'dataarray_numpy':
+            dst_arr = xr.DataArray(arr, dims=dims)
+        else:  # to.lower() == 'dataarray_dask':
+            dst_arr = xr.DataArray(da.from_array(arr), dims=dims)
+        return dst_arr, 'numpy'
+    # Dask
+    elif isinstance(arr, da.Array):
+        if to.lower() == 'numpy':
+            dst_arr = arr.compute()
+        elif to.lower() == 'dask':
+            dst_arr = arr
+        elif to.lower() == 'dataarray_numpy':
+            dst_arr = xr.DataArray(arr.compute(), dims=dims)
+        else:  # to.lower() == 'dataarray_dask':
+            dst_arr = xr.DataArray(arr, dims=dims)
+        return dst_arr, 'dask'
+
+    # DataArray_Numpy
+    elif isinstance(arr, xr.DataArray) and isinstance(arr.data, np.ndarray):
+        if to.lower() == 'numpy':
+            dst_arr = arr.data
+        elif to.lower() == 'dask':
+            dst_arr = da.from_array(arr.data)
+        elif to.lower() == 'dataarray_numpy':
+            dst_arr = arr
+        else:  # to.lower() == 'dataarray_dask':
+            dst_arr = xr.DataArray(da.from_array(arr.data), dims=dims)
+        return dst_arr, 'DataArray_Numpy'
+
+    # DataArray_Dask
+    elif isinstance(arr, xr.DataArray) and isinstance(arr.data, da.Array):
+        if to.lower() == 'numpy':
+            dst_arr = arr.data.compute()
+        elif to.lower() == 'dask':
+            dst_arr = arr.data
+        elif to.lower() == 'dataarray_numpy':
+            dst_arr = arr.compute()
+        else:  # to.lower() == 'dataarray_dask':
+            dst_arr = arr
+        return dst_arr, 'DataArray_Dask'
```

Comment:
Each of these blocks should be its own function.

Reply:
Should I define these functions outside `geometry.py`? In `utils/array.py` maybe?

Reply:
Yes, that definitely doesn't belong here. Problably this whole function should be moved there then.

(Not in the scope of this PR) But I'm wondering, in the grand scheme of things, if it should be preferable to have subclasses for each type of container, like `_SwathDefinition_numpy`, `_SwathDefinition_dask`, etc

Reply:
Moved to utils/array.py and each block separated in its own function.
I wonder what name convention is better: 
- `dataarray_dask` or `xarray_dask` 
- `dataarray_numpy` or `xarray_numpy`

@mraspaud @djhoese  any preference?